core/include/clist.h

```diff
@@ -98,23 +171,175 @@ static inline clist_node_t *clist_remove_head(clist_node_t *list)
+ * @brief Returns first element in list
  *
- * @param[in]       clist       The list to get printed out.
+ * @note: Complexity: O(1)
+ *
+ * @param[in]   list        The list to work upon.
+ * @returns     first (leftmost) list element, or NULL if list is empty
  */
-void clist_print(clist_node_t *clist);
-#endif
+static inline clist_node_t *clist_lpeek(const clist_node_t *list)
+{
+    if (list->next) {
+        return list->next->next;
+    }
+}
+
+/**
+ * @brief Returns last element in list
+ *
+ * @note: Complexity: O(1)
+ *
+ * @param[in]   list        The list to work upon.
+ * @returns     last (rightmost) list element, or NULL if list is empty
+ */
+static inline clist_node_t *clist_rpeek(const clist_node_t *list)
+{
+    return list->next;
+}
+
+/**
+ * @brief Removes and returns last element from list
+ *
+ * @note Complexity: O(n) with n being the number of elements in the list.
+ *
+ * @param[in,out]   list        Pointer to the *list* to remove last element
+ *                              from.
+ */
+static inline clist_node_t *clist_rpop(clist_node_t *list)
+{
+    if (list->next) {
+        list_node_t *last = list->next;
+        while (list->next->next != last) {
+            clist_lpoprpush(list);
+        }
+        return clist_lpop(list);
+    }
+    else {
+        return NULL;
+    }
+}
+
+/**
+ * @brief Finds node and returns its predecessor
+ *
+ * @note Complexity: O(n)
+ *
+ * @param[in]       list    pointer to clist
+ * @param[in,out]   node    Node to look for
+ *                          Must not be NULL.
+ *
+ * @returns         predecessor of node if found
+ * @returns         NULL if node is not a list member
+ */
+static inline clist_node_t *clist_find_before(const clist_node_t *list, const clist_node_t *node)
+{
+    clist_node_t *pos = list->next;
+    if (!pos) {
+        return NULL;
+    }
+    do {
+        pos = pos->next;
+        if (pos->next == node) {
+            return pos;
+        }
+    } while (pos != list->next);
+
+    return NULL;
+}
+
+/**
+ * @brief Finds and returns node
+ *
+ * @note Complexity: O(n)
+ *
+ * @param[in]       list    pointer to clist
+ * @param[in,out]   node    Node to look for
+ *                          Must not be NULL.
+ *
+ * @returns         node if found
+ * @returns         NULL if node is not a list member
+ */
+static inline clist_node_t *clist_find(const clist_node_t *list, const clist_node_t *node)
+{
+    clist_node_t *tmp = clist_find_before(list, node);
+    if (tmp) {
+        return tmp->next;
+    }
+    else {
+        return NULL;
+    }
+}
+
+/**
+ * @brief Finds and removes node
+ *
+ * @note Complexity: O(n)
+ *
+ * @param[in]       list    pointer to clist
+ * @param[in,out]   node    Node to remove for
+ *                          Must not be NULL.
+ *
+ * @returns         node if found and removed
+ * @returns         NULL if node is not a list member
+ */
+static inline clist_node_t *clist_remove(clist_node_t *list, clist_node_t *node)
```

Comment:
Some of them might be. I kept them static inline as they're used by the scheduler, where every instruction counts. I'll do an evaluation of the benefits/downsides in a folo.